File: 실버 2 (1문제)/1012_유기농 배추.c
// 참고 링크 : https://moon-gd.tistory.com/58
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 51

int dir_x[4] = { 0, 1, 0 , -1 };
int dir_y[4] = { -1, 0, 1, 0 };
int ground[MAX][MAX];
int parent[MAX][MAX];
int x_size, y_size, connect_size;

typedef struct node {
	int x;
	int y;
}node;
node stack[MAX * MAX]; 
int parent_index = 0;

void bfs(int x, int y);

int main() {
	int test;
	scanf("%d", &test);

	for (int i = 1; i <= test; i++) {
		scanf("%d %d %d", &x_size, &y_size, &connect_size);
		int x, y;
		parent_index = 1;
		for (int j = 0; j < connect_size; j++) {
			scanf("%d %d", &x, &y);
			ground[x][y] = 1;
			node temp = { x, y };
			stack[j] = temp;
		}

		// 하나씩 꺼내가면서 부모 체크
		for (int j = 0; j < connect_size; j++) {
			x = stack[j].x, y = stack[j].y;
			if (parent[x][y]) {
				continue;
			}
			bfs(x, y);
			parent_index++;
		}

		// 결과 출력
		printf("%d\n", parent_index - 1);

		for (int j = 0; j < y_size; j++) {
			for (int k = 0; k < x_size; k++) {
				ground[k][j] = 0;
				parent[k][j] = 0;
			}
		}
	}


	return 0;
}

void bfs(int x, int y) {
	if (parent[x][y]) {
		return;
	}
	parent[x][y] = parent_index;

	for (int i = 0; i < 4; i++) {
		int new_x = x + dir_x[i];
		int new_y = y + dir_y[i];

		// continue : 유효 범위를 벗어난 경우 1
		if (new_x < 0 || new_y < 0) {
			continue;
		}

		// continue : 유효 범위를 벗어난 경우 2
		if (new_x >= x_size || new_y >= y_size) {
			continue;
		}

		// continue : 이미 부모가 체크된 경우
		if (parent[new_x][new_y]) {
			continue;
		}

		if (ground[new_x][new_y]) {
			bfs(new_x, new_y);
		}
	}
}

File: 실버 2 (8문제)/1012_유기농 배추.py
# 참고 링크 : https://moon-gd.tistory.com/58
import sys
sys.setrecursionlimit(10000)  # 재귀 깊이 10000 설정
q_input = sys.stdin.readline  # 빠른 입력을 받기 위한 준비

testcase = int(q_input())

dir_x = [0, 1, 0, -1]
dir_y = [-1, 0, 1, 0]
parent = []
ground = []
parent_index = 1

for _ in range(51):
    temp_list1 = []
    temp_list2 = []
    for _ in range(51):
        temp_list1.append(0)
        temp_list2.append(0)
    parent.append(temp_list1)
    ground.append(temp_list2)


def bfs(X, Y):
    # 종료 조건 : 이미 소속된 땅이 존재하는 경우
    if parent[X][Y]:
        return
    else:
        parent[X][Y] = parent_index
        for i in range(4):
            new_X = X + dir_x[i]
            new_Y = Y + dir_y[i]

            # continue : 유효 범위를 벗어난 경우 1
            if new_X < 0 or new_X >= x_size:
                continue

            # continue : 유효 범위를 벗어난 경우 2
            if new_Y < 0 or new_Y >= y_size:
                continue

            if ground[new_X][new_Y]:
                bfs(new_X, new_Y)



for _ in range(testcase):
    x_size, y_size, connect_size = map(int, q_input().split())

    stack = []
    parent_index = 1

    # 배추 정보 입력받기
    for _ in range(connect_size):
        x, y = map(int, q_input().split())
        stack.append([x, y])
        ground[x][y] = 1


    for i in range(connect_size):
        x, y = stack[i][0], stack[i][1]

        # continue : 이미 속한 구역이 있을 경우
        if parent[x][y]:
            continue
        bfs(x, y)
        parent_index = parent_index + 1

    print(parent_index - 1)

    # 초기화 과정
    for y in range(y_size):
        for x in range(x_size):
            parent[x][y] = 0
            ground[x][y] = 0
